third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp

```diff
@@ -0,0 +1,201 @@
+                         : schedulingType;
+  }
+
+  enum SchedulingType : uint32_t { NONE = 0, IGLP_0, IGLP_1, COUNT };
+
+  LogicalResult
+  matchAndRewrite(triton::gpu::InstructionBlockSchedOp instructionSchedOp,
+                  PatternRewriter &rewriter) const override {
+
+    bool useSchedulingGuards = true;
+    if ((schedulingType == SchedulingType::IGLP_0) ||
+        (schedulingType == SchedulingType::IGLP_1))
+      useSchedulingGuards = false;
+
+    Operation *op;
+    Block *block = instructionSchedOp->getBlock();
+    rewriter.setInsertionPointToStart(block);
+    if (useSchedulingGuards)
+      generatedSchedBarrier(rewriter, InstructionMaskEnum::NONE);
+
+    rewriter.setInsertionPoint(block, std::prev(block->end()));
+
+    switch (schedulingType) {
+    case SchedulingType::IGLP_0:
+      [[fallthrough]];
+    case SchedulingType::IGLP_1: {
+      generatedIglpOpt(rewriter, static_cast<int>(schedulingType - 1));
+      break;
+    }
+    case SchedulingType::NONE:
+      [[fallthrough]];
+    default: {
+      break;
+    }
+    }
+
+    if (useSchedulingGuards)
+      op = generatedSchedBarrier(rewriter, InstructionMaskEnum::NONE);
+
+    rewriter.eraseOp(instructionSchedOp);
+    return mlir::success();
+  }
+
+private:
+  SchedulingType schedulingType;
+};
+
+struct LowerInstructionSchedOps
+    : public triton::impl::LowerInstructionSchedOpsBase<
+          LowerInstructionSchedOps> {
+
+  explicit LowerInstructionSchedOps(uint32_t variant) {
+    this->variant = variant;
+  }
+
+  void runOnOperation() override {
+    MLIRContext *ctx = &getContext();
+    ModuleOp mod = getOperation();
+
+    ConversionTarget target(*ctx);
+    target.addLegalDialect<LLVM::LLVMDialect>();
+    target.addIllegalOp<triton::gpu::InstructionBlockSchedOp>();
+
+    RewritePatternSet patterns(ctx);
+    patterns.add<InstructionSchedOpRewriter>(ctx, this->variant);
+
+    if (failed(applyPartialConversion(getOperation(), target,
+                                      std::move(patterns)))) {
+      signalPassFailure();
+    }
+  }
+};
+
+struct InsertInstructionSchedOps
+    : public triton::impl::InsertInstructionSchedOpsBase<
+          InsertInstructionSchedOps> {
+
+  void insertPlaceholder(mlir::OpBuilder &builder, triton::DotOp dot) {
+    mlir::OpBuilder::InsertionGuard guard(builder);
+    builder.setInsertionPointAfter(dot);
+    Location loc = builder.getUnknownLoc();
+    builder.create<triton::gpu::InstructionBlockSchedOp>(loc);
+  }
+
+  void runOnOperation() override {
+    MLIRContext *ctx = &getContext();
+    ModuleOp mod = getOperation();
+
+    mlir::OpBuilder builder(ctx);
+    mod.walk(
+        [this, &builder](triton::DotOp op) { insertPlaceholder(builder, op); });
+  }
+};
+} // namespace
+
+namespace mlir {
+namespace triton {
+std::unique_ptr<OperationPass<ModuleOp>>
+createLoweringInstructionSchedOpsPass(uint32_t variant) {
+  return std::make_unique<LowerInstructionSchedOps>(variant);
```

Comment:
this looks like just lowering to LLVM. Should it be part of the pass to convert to LLVM?